nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ExtractEmailAttachments.java

```diff
@@ -0,0 +1,209 @@
+                        MimeMessageParser parser = new MimeMessageParser(originalMessage).parse();
+                        // RFC-2822 determines that a message must have a "From:" header
+                        // if a message lacks the field, it is flagged as invalid
+                        Address[] from = originalMessage.getFrom();
+                        Date sentDate = originalMessage.getSentDate();
+                        if (from == null || sentDate == null ) {
+                            // Throws MessageException due to lack of minimum required headers
+                            throw new MessagingException("Message failed RFC2822 validation");
+                        }
+                        if (parser.hasAttachments()) {
+                            final String originalFlowFileName = originalFlowFile.getAttribute(CoreAttributes.FILENAME.key());
+                            for (final DataSource data : parser.getAttachmentList()) {
+                                try {
+                                FlowFile split = session.create(originalFlowFile);
+                                final Map<String, String> attributes = new HashMap<>();
+                                if (StringUtils.isNotBlank(data.getName())) {
+                                    attributes.put(CoreAttributes.FILENAME.key(), data.getName());
+                                }
+                                if (StringUtils.isNotBlank(data.getContentType())) {
+                                    attributes.put(CoreAttributes.MIME_TYPE.key(), data.getContentType());
+                                }
+                                String parentUuid = originalFlowFile.getAttribute(CoreAttributes.UUID.key());
+                                attributes.put(ATTACHMENT_ORIGINAL_UUID, parentUuid);
+                                attributes.put(ATTACHMENT_ORIGINAL_FILENAME, originalFlowFileName);
+                                split = session.append(split, new OutputStreamCallback() {
+                                    @Override
+                                    public void process(OutputStream out) throws IOException {
+                                        IOUtils.copy(data.getInputStream(),out);
+                                    }
+                                });
+                                split = session.putAllAttributes(split, attributes);
+                                originalFlowFilesList.add(originalFlowFile);
+                                attachmentsList.add(split);
+                            } catch (FlowFileHandlingException e) {
+                                    session.remove(attachmentsList);
+                                    logger.error("Flowfile {} triggered error {} while processing message removing generated FlowFiles from sessions", new Object[] {originalFlowFile, e});
+                                    invalidFlowFilesList.add(originalFlowFile);
+                                }
```

Comment:
If there are multiple attachments and one fails, the originalFlowFile will get added to the invalid list and success list. In the interest of simplifying the processor a bit, probably best to put this try/catch around the for loop so you don't have to handle the cases where some attachments fail another others succeed. 

If you decided to handle that, you need a way to convey to the user which attachments failed, which succeeded, where the original is transferred, etc.